File: php/src/data-structures/LinkedList.php
<?php

namespace DataStructures;

use DataStructures\Node;

class LinkedList
{
    /**
     * @var int
     */
    private $size;
    /**
     * @var Node|null
     */
    private $head;

    /**
     * LinkedList constructor.
     */
    public function __construct()
    {
        $this->size = 0;
        $this->head = null;
    }

    /**
     * @param mixed $data
     */
    public function addFront($data): void
    {
        $node = new Node($data);

        //For the first data
        if ($this->head != null) {
            $node->next = $this->head;
        }

        //Head is now the new node
        $this->head = $node;
        $this->increaseSize();
    }

    /**
     * @param mixed $data
     */
    public function addBack($data): void
    {
        $newTail = new Node($data);

        $tail = $this->goToTail();

        if ($tail != null) {
            $tail->next = $newTail;
        } else {
            $this->head = $newTail;
        }

        $this->increaseSize();
    }

    /**
     * @return int
     */
    public function size(): int
    {
        return $this->size;
    }

    /**
     * @return mixed|null
     */
    public function getFirst()
    {
        //Not identical
        if ($this->head !== null) {
            return $this->head->getData();
        }
        return null;
    }

    /**
     * @return mixed|null
     */
    public function getLast()
    {
        //Cannot call method getData() on DataStructures\Node|null.
        //return $this->goToTail()->getData() ?? null;
        // PHPStan problem : Cannot call method getData() on DataStructures\Node|null.
        //https://github.com/phpstan/phpstan/issues/1157

        //Not identical
        if ($this->goToTail() !== null) {
            return $this->goToTail()->getData();
        }
        return null;
    }

    /**
     * @param mixed $value
     */
    public function delete($value): void
    {
        if ($this->head == null) {
            return;
        }

        $current = $this->head;

        //if head
        if ($current->getData() == $value) {
            $this->head = $current->next;
            $this->size--;
            return;
        }

        while ($current->next != null) {
            if ($current->next->getData() == $value) {
                $current->next = $current->next->next;
                $this->size--;
                break;
            }
            $current = $current->next;
        }
    }

    private function increaseSize(): void
    {
        $this->size++;
    }

    private function goToTail(): ?Node
    {
        $node = $this->head;

        while ($node != null && $node->next != null) {
            $node = $node->next;
        }

        return $node;
    }
}

File: ruby/lib/data-structures/linked_list.rb
class LinkedList
  attr_reader :size

  def initialize
    @size = 0
    @head = nil
  end

  def add_front(value)
    new_node = Node.new(value)
    new_node.next = @head unless @head.nil?

    @head = new_node
    increase_size
  end

  def add_back(value)
    node = Node.new(value)

    if @head.nil?
      @head = node
      increase_size
      return
    end

    tail = to_tail

    if tail.nil?
      @head.next = node
    else
      tail.next = node
    end
    increase_size
  end

  def first
    @head ? @head.data : nil
  end

  def last
    tail = to_tail

    tail ? tail.data : nil
  end

  def delete(value)
    return if @head.nil?

    current = @head

    if value == current.data
      @head = current.next
    else
      while !current.next.nil?
        if value == current.next.data
          current.next = current.next.next
          break
        end
        current = current.next
      end
    end
    decrease_size
  end

  def to_s
    node = @head
    puts node
    while (node = node.next)
      puts node
    end
  end

  private

  def to_tail
    current = @head
    current = current.next until current.next.nil?
    current
  end

  def increase_size
    @size += 1
  end

  def decrease_size
    @size -= 1
  end

  class Node
    attr_reader :data
    attr_accessor :next

    def initialize(data)
      @data = data
      @next = nil
    end

    def to_s
      "Node [data: #{@data}]"
    end
  end
end
